### PR TITLE
Feature flag improvements

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Feature

## Description

**The good**

* We use feature flags 🎉 

**The bad**

* Creating a feature flag involves manual work, like creating a data update script (DUS) which calls `FeatuerFlag.add(...)`.
* There's no place in the codebase where one can see all feature flags, only the Flipper web UI or a DB query.

**The ugly**

* Method calls like `FeatureFlag.add`, `FeatureFlag.enabled?`, etc. work with strings or symbols, so it's easy for typos to sneak in.

**Proposed solution in this PR**

* A new config file (`config/feature_flags.yml`) that lists all available feature flags (so far I added just one for demo purposes). This is the canonical list of all feature flags used in the app. This does NOT need to be manually managed.
* A new generator. Calling `rails generate feature_flag test` will do the following:
    1. It appends a line with the new flag to the YAML list in `config/feature_flags.yml`.
    2. It creates the DUS for adding the feature flag automatically. To avoid duplication of effort this re-uses the existing DUS generator.
* A new block in `application.rb`'s `after_intialize` section that define a predicate method for each new feature flag, e.g. `FeatuereFlag.profile_admin_enabled?`. It also defines a constant (e.g. `FeatureFlag::FLAG_PROFILE_ADMIN`) which can be used wherever the actual flag name as string is required.

Here's the generator in action:

```
❯ rails generate feature_flag test
      append  config/feature_flags.yml
    generate  data_update
       rails  generate data_update add_test_feature_flag --no-spec
      create  lib/data_update_scripts/20220204061126_add_test_feature_flag.rb
        gsub  lib/data_update_scripts/20220204061126_add_test_feature_flag.rb
``` 

And the generated DUS:

```ruby
module DataUpdateScripts
  class AddTestFeatureFlag
    def run
      FeatureFlag.add(:test)
    end
  end
end
```

To illustrate how the changes affect the codebase I updated code related to the `profile_admin` and `data_update_scripts` flags to make use of the constants and predicate methods.

## QA Instructions, Screenshots, Recordings

I mostly QA'd this manually, i.e.run the generator and look at the generated/modified artifacts.

### UI accessibility concerns?

No UI changes

## Added/updated tests?

- [X] No, and this is why: we don't generally test our custom generators. Maybe this should change at one point but I'd like to keep that for a future PR.

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

## [optional] What gif best describes this PR or how it makes you feel?

fun with flags

### CODE REVIEW

File: lib/generators/feature_flag/USAGE [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Description:
```
Comment (citizen428) [inline lib/generators/feature_flag/USAGE:1]: I didn't bother changing this yet. If we go ahead with this proposed solution I'll obviously do that.

File: lib/generators/feature_flag/feature_flag_generator.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+class FeatureFlagGenerator < Rails::Generators::NamedBase
+  REPLACEMENT_REGEX = /(def run).*?(end.*)\z/m
+
+  def create_feature_flag_config
+    append_file "config/feature_flags.yml", %(  - #{file_name}\n)
+  end
+
+  def create_dus
+    script_name = "add_#{file_name}_feature_flag"
+    generate "data_update", script_name, "--no-spec"
```
Comment (citizen428) [inline lib/generators/feature_flag/feature_flag_generator.rb:10]: Not generating a spec seemed sensible here, we'd essentially just be testing that `FeatureFlag.add` works which we do elsewhere.

File: lib/generators/feature_flag/feature_flag_generator.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    file = Dir.glob("lib/data_update_scripts/*#{script_name}.rb").first
```
Comment (citizen428) [inline lib/generators/feature_flag/feature_flag_generator.rb:11]: This is not super robust but IMO good enough™. Unless someone manually creates scripts with conflicting names we should be ok.

File: lib/generators/feature_flag/feature_flag_generator.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    raise "Can't find DataUpdateScript" unless file
```
Comment (citizen428) [inline lib/generators/feature_flag/feature_flag_generator.rb:12]: We could add another condition that checks if there's more than one existing script now because that probably means someone accidentally ran the generator twice. Though I think that's easy enough to catch while preparing the PR.

**Comment by jeremyf** [comment]
I like where this is heading.  I love the idea of using the `yml` file as something we can reference through `git annotate` (e.g. see what all features we've been enabling and disabling).

I'd be partial to having `FeatureFlag.creator_onboarding_enabled?` and `FeatureFlag.creator_onboarding_disabled?` be two additional methods added `after_initialize`.  The reason, it looks a bit odd seeing `FeatureFlag.___(FeatureFlag.___)`; almost like `FeatureFlag` is having a conversation with itself.  Not bad, just a bit more chatty.

**Comment by JennieOcken** [comment]
I have 0 comments on the implementation but I am utterly in love with the idea.  I have lived the horrors of having a feature flag accidently flipped or incorrectly setup in the first place and half baked functionality going live.  Also, this would help me, from a product standpoint, get a better understand about what we have in place but not "core" to the application (in those cases where something is feature flagged not because we are in active development but because we decide only certain Forems need/get it, e.g. Listings).  Huzzah.

**Comment by maestromac** [comment]
Does the custom generator automatically support `rails destroy feature_flag test`? It's not important I'm just curious

**Comment by joshpuetz** [comment]
Absolutely love this idea!

**Comment by citizen428** [comment]
@jeremyf Agreed, it feels a bit awkward. I actually started with dynamically defining constants (e.g. `FLAG_CREATOR_ONBOARDING`) so the code would have been `FeatureFlag.enabled?(FLAG_CREATOR_ONBOARDING)`.  This feels better from a code perspective but I don't like messing with constants in this way because of Zeitwerk, autoloading, etc.

I do like `FeatureFlag.creator_onboarding_enabled?` but I don't think I'm going to add `FeatureFlag.creator_onboarding_disabled?` as this can be quite easily achieved by using `unless` in the condition.

**Comment by citizen428** [comment]
@maestromac Sort of. I twill remove the line from `config/feature_flags.yml` (I guess that only works as long as it's still the last line) but won't remove the DUS for reasons I'm not quite willing to investigate right now 😉

**Comment by citizen428** [comment]
One interesting issue: the constants defined in `application.rb` seem to work fine everywhere except for the "Front-end tests, Storybook, bundle audit, console check" runner on Travis.

File: config/application.rb [hunk 5]
```diff
@@ -67,9 +67,8 @@ class Application < Rails::Application
     # Globally handle Pundit::NotAuthorizedError by serving 404
     config.action_dispatch.rescue_responses["Pundit::NotAuthorizedError"] = :not_found
 
-    # After-initialize checker to add routes to reserved words
```
Comment (citizen428) [inline config/application.rb:70]: I'm not only using a single `after_initialize` block so the comment doesn't make sense here anymore. It was a bit redundant with the one 2 lines below anyway.

**Comment by citizen428** [comment]
Sorry for the force push, I rebased with `main` to suppress the warnings related to `device_selector`.

> One interesting issue: the constants defined in `application.rb` seem to work fine everywhere except for the "Front-end tests, Storybook, bundle audit, console check" runner on Travis.

So, this was fun. I had changed  `app/models/admin_menu.rb` to contain the following constant:

```ruby
FEATURE_FLAGS = [
  FeatureFlags::PROFILE_ADMIN,
  FeatureFlags::DATA_UPDATE_SCRIPTS,
].freeze
```

Because constant definitions happen in the class body these get evaluated when the class is read and you know what didn't run at this point? Correct, the `config.after_initialize` code block in `config/application.rb`. This only happened on one specific Travis worker because that one runs in the `production` environment where `config.eager_load = true`.  Yes, that took me a moment to wrap my head around. 

For now, I change the constant to include the predicate methods and use them for checks further down the file, which I think is good enough for now. I probably want to look at the code again at a later point as we're checking each flag twice right now, once in `navigtation_items` and once in `feature_flagged_menu_items`. I guess the reason for this is avoiding `dup`ing the relatively big `ITEMS` constant but we do so at the expense of extra DB lookups. Anyway, this is a change for another time.

I also keep flip-floping on whether or not I want to use constants or methods for the flag "accessors". Then I realized that the code changes in the current PR need neither, only the predicate methods. So I decided to do neither for now, should we ever really need something like this will have actual usages to better inform this decision.

**Comment by citizen428** [comment]
tl;dr I was just trying to address things that various people brought up during conversations and PR discussions. You seem to feel strongly about this topic and its implications for SRE work, so if you volunteer to spearhead the effort in reducing friction around feature flags I'll more than happily close this PR.

---

> If nothing else, the product team would know because they would typically control the launch of features.

Based on Jennie's component above, especially the product team was in favor of this. 

I guess overall I agree, we're (still) a small company. In my previous job, several hundred engineers across 30 teams worked on hundreds of microservices. I don't think there was anyone who would have been able to tell you all the feature flags in use.

> Can you talk about the need to create feature flags? Is simply toggling them insufficient?

Maybe not a need but something we aligned around as a team early on when adding Flipper. IIRC rhymes was a big proponent of this because he didn't want to rely on implicit behavior. I'm sure we could unearth the related discussion from past PRs. I have no strong feelings on this, primarily because of Flipper's stringly-typed API. If `Flipper.enabled?(:falg)` would e.g. raise if no flag with this name exists instead of just returning `false` it'd be a different story. 

> This is not accurate. A search under the app directory for FeatureFlag.enabled?

Sorry, but IMO "you can always just use `grep`" is not a valid argument. It's a bit like saying you don't need a curated changelog, you can just grep through `git log`.

There are people who aren't developers who may not be able to do that. `grep` also doesn't easily provide historical context as to when specific features were en- or disabled.

> Also, can you expand on how the web UI 

Primarily just context switching. I also wasn't 100% sure if everyone on the team (including all non-US team members) has access to all routes under `/admin`. If they do, it's a good way to look at the current state, again without historical context.

> or running Flipper.features in a Rails console are insufficient

`Flipper.features` is good, I think I wasn't aware of that and if I was I forgot about it 😅  but IMO we should add that to the forwards in our `FeatureFlag` wrapper class and I'll open a PR for that. Thanks!

> Didn't require a framework with monkeypatching, just constants. 

I guess this depends on your definition of monkey-patching. IMO not every dynamic method definition is a monkey-patch. But point taken.

**Comment by JennieOcken** [comment]
> > If nothing else, the product team would know because they would typically control the launch of features.
> 
> Based on Jennie's comment above, especially the product team was in favor of this.

Reiterating that I have 0 opinion on technical implementation but, I was invoked so I can state an opinion on feature flags.  😄 .

I am seeing three major use cases of feature flags here:

1) To incrementally ship pieces of a large feature and then turn the entire feature on when all the pieces are in production.  I have most seen this on the Admin Experience team.  It worked pretty slick because it allowed us to time the release without creating a long running branch or dealing with a million last minute merge conflicts pushing a big feature back into the monolith.  I have come around that this may be a better/safer way to do releases (a previous gig used GitFlow, a lot of branches, and occasionally failed feature flags in .net and there were some painful moments).  However, the risk we need to mitigate here is ensuring we flip all the feature flags on release.  Relying on the developers on the team to manually track or, worse, remember all the feature flags makes me nervous.

2) A light weight framework for selectively deactivating features for certain installations.  This is what we are doing on Content Experience team right now with Listings.  Eventually we may have a more complete process for this called a Plugin Marketplace but for now this will allow us to define certain features as "core" to a Forem and remove features that may be unnecessary or in some cases problematic.  I have lots of Creators who are confused by Listings and will never use them.  I have one Creator whose members would be offended at having any trace of them in the application (very concerned about being monetized).  This is the place where a list of flags would be extremely helpful to me as I am working on defining our "core" versus "add on" functionality plus some cut lists.  For example, we may make Listings only available in dev.to. 

3) To create a clean fall back for a feature release when a bug is found.  I saw this happen most recently with some feed experiments.  This is an area where I take off my product hat (because Product don't dabble here) and put on my ??Leadership?? hat.  If we have a production failure, I want to have it resolved quickly.  I think how we do that ends up being a lot of collaboration between SRE, QA, and the developers.  Having lived through a lot of firefighting I know how the brain doesn't work as well when under pressure.  I don't want to rely on memory at those points.  We need to have clean rules and clear paths.  We probably need to keep thinking through the best way to prepare for and resolve these issues (including preventing them in the first place) but I think a cleaner process for this that doesn't rely on discovering information is generally a good thing.

So, that's the good word from ProductLand.  We are going to keep using feature flags.  We are likely going to use them more.  I think it is a good idea for us to figure out a way to use them consistently.  I trust the people engaged in this conversation to come to a good solution that works across development and SRE and is for the best of the company as a whole.

**Comment by citizen428** [comment]
> Reiterating that I have 0 opinion on technical implementation 

💯 , bad phrasing on my part, "in favor of this" somewhat implied this concrete implementation when instead I should have said "in favor of something like this".

**Comment by citizen428** [comment]
Didn't feel called out at all, no worries. I just realized that I quoted you in a way that was a bit disingenuous as it was bolstering my own argument and while that was subconscious it was still wrong.
